provider/slack/client.ts

```diff
@@ -0,0 +1,311 @@
+import type { ItemsResult } from '@openctx/provider'
+import { WebClient } from '@slack/web-api'
+import type { MessageElement } from '@slack/web-api/dist/types/response/ConversationsHistoryResponse.js'
+import type { Match } from '@slack/web-api/dist/types/response/SearchFilesResponse.js'
+import dedent from 'dedent'
+import { XMLBuilder } from 'fast-xml-parser'
+
+const xmlBuilder = new XMLBuilder({ format: true })
+
+interface ChannelInfo {
+    name: string
+    id: string
+    url: string
+}
+
+export class SlackClient {
+    private readonly client: WebClient
+    public channelList: ChannelInfo[] = []
+    public useridMapping: Map<string, string> = new Map()
+
+    constructor(slackAuthToken: string) {
+        this.client = new WebClient(slackAuthToken)
+    }
+
+    public getChannelList(): ChannelInfo[] {
+        return this.channelList
+    }
+
+    public async contextCandidatesFromRecentThreads(
+        channelId: string | undefined,
+        limit = 10
+    ): Promise<ItemsResult> {
+        if (!channelId) {
+            return []
+        }
+        const conversations = await this.client.conversations.history({
+            channel: channelId,
+            limit,
+        })
+        const messages = conversations.messages || []
+        messages.sort((a, b) => (b.ts ?? '').localeCompare(a.ts ?? ''))
+
+        const allPromises = messages.map(async msg => {
+            const threadTs = msg.ts
+            if (threadTs === undefined) {
+                return undefined
+            }
+            const threadConversation = await this.fetchThreadMessages(channelId, threadTs)
+            // Basic checks to ensure some minimum conversation in the thread
+            if (!threadConversation || threadConversation?.trim().length < 300) {
+                return undefined
+            }
```

Comment:
how did you come up with 300? Did you see poor results on threads with less characters? Just wondering, what if there is a 50 character highly relevant reply - then it would be silently ignored.

Reply:
From the initial experiment, 300 characters were picking short messages and mostly seemed non-relevant by hit and trial experiments, but I definitely agree that sometimes replies can highly relevant even with less characters. I think this can be more experimental and would makes sense once we get some feedback on the provider itself. wdyt ?